source/MaterialXGenShader/ShaderGraph.cpp

```diff
@@ -843,35 +670,47 @@ ShaderNode* ShaderGraph::createNode(const Node& node, GenContext& context)
+    const string& targetColorSpace = context.getOptions().targetColorSpaceOverride.empty() ?
+                                     _document->getActiveColorSpace() :
+                                     context.getOptions().targetColorSpaceOverride;
     const string& targetDistanceUnit = context.getOptions().targetDistanceUnit;
 
-    ColorManagementSystemPtr colorManagementSystem = context.getShaderGenerator().getColorManagementSystem();
-    string targetColorSpace = context.getOptions().targetColorSpaceOverride.empty() ?
-                              _document->getActiveColorSpace() :
-                              context.getOptions().targetColorSpaceOverride;
-
-    for (InputPtr input : node.getInputs())
+    for (InputPtr input : node->getInputs())
     {
-        if (input->getType() == FILENAME_TYPE_STRING)
+        if (input->hasValue() || input->hasInterfaceName())
         {
-            ShaderOutput* shaderOutput = newNode->getOutput();
-            if (shaderOutput)
+            string sourceColorSpace = input->getActiveColorSpace();
+            if (input->getType() == FILENAME_TYPE_STRING && node->isColorType())
             {
-                string colorSpace = populateColorTransformMap(colorManagementSystem, shaderOutput, input, targetColorSpace, false);
-                ShaderInput* shaderInput = newNode->getInput(input->getName());
-                if (shaderInput && !colorSpace.empty())
+                // Adjust the source color space for filename interface names.
+                if (input->hasInterfaceName())
                 {
-                    shaderInput->setColorSpace(colorSpace);
+                    for (ConstNodePtr parentNode : context.getParentNodes())
+                    {
+                        if (!parentNode->isColorType())
+                        {
+                            InputPtr interfaceInput = parentNode->getInput(input->getInterfaceName());
+                            string interfaceColorSpace = interfaceInput ? interfaceInput->getActiveColorSpace() : EMPTY_STRING;
+                            if (!interfaceColorSpace.empty())
+                            {
+                                sourceColorSpace = interfaceColorSpace;
+                            }
+                        }
+                    }
                 }
+
+                ShaderOutput* shaderOutput = newNode->getOutput();
+                populateColorTransformMap(colorManagementSystem, shaderOutput, sourceColorSpace, targetColorSpace, false);
                 populateUnitTransformMap(unitSystem, shaderOutput, input, targetDistanceUnit, false);
             }
-        }
-        else
-        {
-            ShaderInput* shaderInput = newNode->getInput(input->getName());
-            populateColorTransformMap(colorManagementSystem, shaderInput, input, targetColorSpace, true);
-            populateUnitTransformMap(unitSystem, shaderInput, input, targetDistanceUnit, true);
+            else
```

Comment:
I think this logic is inconsistent? i.e. interface inputs are checked for filename but not for non-filenames ?

Reply:
This is a great question, and the reason that this change is only required for filename inputs has to do with the unusual way that we process color transforms for filename inputs, applying them to the **output** of the associated node.

Imagine the common case of an image node, where the colorspace is assigned to its file input, but the associated transform must be applied to the output of the image node, not to the literal filename-type input.

For this reason, our usual handling of interface names doesn't automatically apply to color space bindings for filename inputs, and we must manually handle this case in the logic of our color transform system.